src/layoutBuilder.js

```diff
@@ -388,75 +495,288 @@ function convertWordsToCodepoints(lineAsArrayOfCodepoints) {
+	line.inlines.forEach((word, index, array) => {
+		// extract the parts we're interested in, and to avoid whitespace woes we will
+		// key using the trimmed text
+		const isLastWord = index === array.length - 1;
+		const lastCharWasSpace = word.text.slice(-1) === ' ';
+		const text = word.text.trim();
+		const font = word.font.name.split('-')[0].replace('TCRegular', '');
+		const style = word.style;
+		const alignment = word.alignment;
+
+		if (font === 'NotoSansCJK') {
+			// for CJK add the text to the buffer
+			CJKbuffer.text += text;
+			CJKbuffer.font = font;
+			CJKbuffer.style = style;
+			// if we hit a space we need to add that to the lookup to deal with different
+			// adjacent CJK languages in the line
+			if (lastCharWasSpace || isLastWord) {
+				wordPropsLookup[CJKbuffer.text] = {
+					font: CJKbuffer.font,
+					style: CJKbuffer.style,
+					alignment: CJKbuffer.alignment,
+				};
+				CJKbuffer.text = '';
 			}
-			return line.addInline(newInline);
+		} else {
+			// non-CJK words are broken up as we want for both LTR and RTL languages
+			wordPropsLookup[text] = { font, style, alignment };
+		}
+	});
+
+	// Turn the inlines into a single string
+	const lineElementsAsString = line.inlines.map((e) => e.text).join('');
+
+	// Use that string to create a bidi instance
+	const bidiString = bidi
+		.from_string(lineElementsAsString, { direction: 'LTR' })
+		.reorder_visually();
+
+	// The bidi instance contains string_arr, which is the original string represented
+	// by a single array of codepoints. Call convertWordsToCodepoints in order to
+	// get back an array (representing the line) of arrays (representing the words) of
+	// codepoints (representing the individual characters)
+	const wordsAsArraysOfCodePoints = convertWordsToInlineCodepoints(
+		bidiString.string_arr
+	);
+	const arrayOfTransformedWords = [];
+
+	// Copied from transformLineForRtl, this fills our arrayOfTransformedWords
+	wordsAsArraysOfCodePoints.forEach((wordAsCodePoints) => {
+		const direction = getWordDirection(wordAsCodePoints);
+		const wordAsString = spreadify(
+			String.fromCharCode,
+			String
+		)(wordAsCodePoints);
+		const bidiWord = bidi
+			.from_string(wordAsString, { direction })
+			.reorder_visually()
+			.toString();
+
+		// although we now have the correct word, we've stripped it of the styling information
+		// passed from the digest-pdf-export process, so now we use the lookup to add that back
+		// noting that we have keyed by trimmed words
+		const originalTextNode = wordPropsLookup[bidiWord.trim()] || {};
+		const newTextNode = Object.assign(originalTextNode, {
+			text: bidiWord,
 		});
+		arrayOfTransformedWords.push(newTextNode);
+	});
+
+	// the current implementation still struggles with spacing between RTL to LTR language transitions,
+	// so manually apped a space in these cases before rebuilding the inlines
+	arrayOfTransformedWords.forEach((word, index, array) => {
+		if (
+			array[index - 1] &&
```

Comment:
Why check the previous index (index - 1)?

Also, I think having a `map` and reassigning `arrayOfTransformedWords` would be a better choice as it will be clearer that mutations are happening here.

Reply:
`array[index - 1]` will not work on the first iteration (will always return undefined I think)

Reply:
Yep - I wanted it to skip out in that case and not do what was inside the if block - this has been superseded by a refactor anyhow.